### PR TITLE
Add evaluation to alps_submitter

### PR DESCRIPTION
**Pull request by uuleon**
The major work of this PR contains the following:
1. refactoring `alpsFiller` to make it adaptive to train and predict
2. add `Get` method which supports default value to make code generation easier
3. support `eval_spec` setting and train_with_evaluation

### CODE REVIEW

**Comment by hungry1526** [comment]
@uuleon It would be a good practice to add some detail information in description sector for a particular PR.

**Comment by uuleon** [comment]
@hungry1526 thanks for the reminder,  the description has been updated.

File: sql/alps_submitter.go [hunk 1]
```diff
@@ -22,16 +22,35 @@ const (
...
+	FeatureColumnCode    string
+	TrainSpec            collection
+	EvalSpec             collection
+
+	// Config
+	OdpsConf    collection
+	DatasetConf collection
+}
+
+type collection struct {
```
Comment (tonyyang-svail) [inline sql/alps_submitter.go]: Not sure we need to add this level of abstraction: `OdpsConf.Dict["Key"]` vs `OdpsConf["Key"]`. I think the latter is more concise. If we need to define a method, such as `Get(key, defaultValue string)`, on `map[string]string`, we can simply give `map[string]string` a name. For example,

```golang
package main

import "fmt"

type collection map[string]string

type collections struct {
	name string
	c2 collection
}

func (c collection) Get(key, fallback string) string {
	if v, ok := c[key]; ok {
		return v
	}
	return fallback
}

func main() {
	config := make(map[string]string)
	config["k1"] = "v1"
	var c collection = config
	fmt.Println(c.Get("k1", "fallback1")) // v1
	fmt.Println(c.Get("k2", "fallback2")) // fallback2
	cs := collections{name: "hi", c2: config}
	fmt.Println(cs) // {hi map[k1:v1]}
}
```
By doing so we can avoid the redundant `Dict` field.

File: sql/alps_submitter.go [hunk 2]
```diff
@@ -441,59 +487,62 @@ func newALPSTrainFiller(pr *extendedSelect) (*alpsFiller, error) {
...
-	dir, err := ioutil.TempDir("/tmp", "alps_scratch_dir")
-	if err != nil {
-		return nil, err
-	}
-
 	return &alpsFiller{
-		TableName:            tableName,
+		IsTraining:           true,
+		TrainInputTable:      tableName,
+		EvalInputTable:       tableName,
```
Comment (tonyyang-svail) [inline sql/alps_submitter.go]: Add `//FIXME(uuleon): Train and Eval should use different dataset.`

File: sql/alps_submitter.go [hunk 3]
```diff
@@ -376,10 +402,21 @@ func generateEstimatorCreator(estimator string, attrs []*attribute) (string, err
...
+	if args != nil {
+		for _, arg := range args {
+			cl = append(cl, arg)
+		}
+	}
 	return fmt.Sprintf("tf.estimator.%s(%s)", estimator, strings.Join(cl, ",")), nil
 }
 
 func newALPSTrainFiller(pr *extendedSelect) (*alpsFiller, error) {
+	scratchDir, err := ioutil.TempDir("/tmp", "alps_scratch_dir_")
```
Comment (tonyyang-svail) [inline sql/alps_submitter.go:413]: Should we remove `scratchDir` after the training?

File: sql/alps_submitter.go (hunk 1 above)
Comment (uuleon) [reply sql/alps_submitter.go]: That is a great idea! Thanks for pointing it out.

File: sql/alps_submitter.go (hunk 3 above)
Comment (uuleon) [reply sql/alps_submitter.go:413]: The `scratchDir` will be deleted until the output model is saved using `sqlfs` which is blocking by `ODPS DB`.

Comment (tonyyang-svail) [reply sql/alps_submitter.go:413]: Would you please at least add a TODO?

Comment (uuleon) [reply sql/alps_submitter.go:413]: Ok, thanks

File: sql/alps_submitter.go [hunk 4]
```diff
@@ -367,7 +391,7 @@ func generateFeatureColumnCode(fcs []interface{}) (string, error) {
 	return fmt.Sprintf("[%s]", strings.Join(codes, ",")), nil
 }
 
-func generateEstimatorCreator(estimator string, attrs []*attribute) (string, error) {
+func generateEstimatorCreator(estimator string, attrs []*attribute, args []string) (string, error) {
```
Comment (tonyyang-svail) [inline sql/alps_submitter.go:394]: Could you please provide an example on `args []string`?

**Review by tonyyang-svail — APPROVED** [review]
LGTM. Great job!

**Pull request merged**